File: src/main/java/gg/generations/rarecandy/tools/gui/PassChain.java
package gg.generations.rarecandy.tools.gui;

import gg.generations.rarecandy.renderer.textures.FrameBuffer;
import gg.generations.rarecandy.renderer.textures.ITexture;
import org.lwjgl.opengl.GL11C;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class PassChain implements AutoCloseable {
    private final List<FullscreenPass> passes = new ArrayList<>();

    private final FrameBuffer a;
    private final FrameBuffer b;

    private int width;
    private int height;

    private final Map<String, FrameBuffer> named = new HashMap<>();
    private final Map<String, Float> namedScales = new HashMap<>();

    public PassChain(int width, int height) {
        this.width = Math.max(1, width);
        this.height = Math.max(1, height);

        a = FrameBuffer.soloAlbedo(this.width, this.height);
        b = FrameBuffer.soloAlbedo(this.width, this.height);
    }

    private static int scaled(int value, float scale) {
        return Math.max(1, Math.round(value * scale));
    }

    public PassChain add(FullscreenPass pass) {
        String key = pass.getOutput();

        if (named.containsKey(key)) {
            throw new IllegalArgumentException("A pass with the key '" + key + "' already exists; each one must be unique.");
        }

        namedScales.put(key, pass.getScale());
        named.put(key, FrameBuffer.soloAlbedo(scaled(width, pass.getScale()), scaled(height, pass.getScale())));

        passes.add(pass);
        return this;
    }


    public void resize(int width, int height) {
        this.width = width;
        this.height = height;

        a.resize(Math.max(1, width), Math.max(1, height));
        b.resize(Math.max(1, width), Math.max(1, height));

        named.forEach((key, fb) -> {
            float scale = namedScales.get(key);
            fb.resize(scaled(width, scale), scaled(height, scale));
        });
    }

    public void reload() {
        passes.forEach(FullscreenPass::reload);
    }

    public void render(FrameBuffer gbuffer, int width, int height) {
        List<FullscreenPass> active = passes.stream().filter(FullscreenPass::isEnabled).toList();

        if (active.isEmpty()) {
            gbuffer.blitToScreen(0, width, height, GL11C.GL_COLOR_BUFFER_BIT, GL11C.GL_NEAREST);
            return;
        }

        FrameBuffer previous = gbuffer;

        for (var pass : active) {
            FrameBuffer target = pass.getOutput() != null ? named.get(pass.getOutput()) : previous == a ? b : a;

            pass.render(gbuffer, previous, named, target);
            previous = target;
        }

        active.getLast().renderToScreen(gbuffer, previous, named, width, height);
    }

    @Override
    public void close() {
        passes.forEach(FullscreenPass::destroy);
        a.close();
        b.close();
        named.values().forEach(FrameBuffer::close);
        named.clear();
    }
}

File: src/main/java/gg/generations/rarecandy/tools/gui/PassChain.kt
package gg.generations.rarecandy.tools.gui

import gg.generations.rarecandy.renderer.textures.framebuffer.FrameBuffer
import gg.generations.rarecandy.renderer.textures.framebuffer.FrameBuffer.Companion.soloAlbedo
import org.lwjgl.opengl.GL11C
import java.lang.AutoCloseable
import java.util.function.Consumer
import kotlin.math.max

class PassChain(width: Int, height: Int) : AutoCloseable {
    private val passes: MutableList<FullscreenPass> = mutableListOf()

    private val a: FrameBuffer
    private val b: FrameBuffer

    private var width: Int
    private var height: Int

    private val named = HashMap<String, FrameBuffer>()
    private val namedScales = HashMap<String, Float?>()

    init {
        this.width = max(1, width)
        this.height = max(1, height)

        a = soloAlbedo(this.width, this.height)
        b = soloAlbedo(this.width, this.height)
    }

    fun add(pass: FullscreenPass): PassChain {
        val key = pass.output!!

        require(!named.containsKey(key)) { "A pass with the key '" + key + "' already exists; each one must be unique." }

        namedScales.put(key, pass.scale)
        named.put(key, soloAlbedo(scaled(width, pass.scale), scaled(height, pass.scale)))

        passes.add(pass)
        return this
    }


    fun resize(width: Int, height: Int) {
        this.width = width
        this.height = height

        a.resize(max(1, width), max(1, height))
        b.resize(max(1, width), max(1, height))

        named.forEach { (key: String?, fb: FrameBuffer?) ->
            val scale: Float = namedScales.get(key)!!
            fb!!.resize(scaled(width, scale), scaled(height, scale))
        }
    }

    fun reload() {
        passes.forEach { it.reload() }
    }

    fun render(gbuffer: FrameBuffer, width: Int, height: Int) {
        val active = passes.stream().filter{ it.enabled.invoke() }.toList()

        if (active.isEmpty()) {
            gbuffer.blitToScreen(0, width, height, GL11C.GL_COLOR_BUFFER_BIT, GL11C.GL_NEAREST)
            return
        }

        var previous = gbuffer

        for (pass in active) {
            val target: FrameBuffer = (if (pass.output != null) named[pass.output] else if (previous == a) b else a)!!

            pass.render(gbuffer, previous, named, target)
            previous = target
        }

        active.last().renderToScreen(gbuffer, previous, named, width, height)
    }

    override fun close() {
        passes.forEach(FullscreenPass::destroy)
        a.close()
        b.close()
        named.values.forEach(Consumer { obj: FrameBuffer? -> obj!!.close() })
        named.clear()
    }

    companion object {
        private fun scaled(value: Int, scale: Float): Int {
            return max(1, Math.round(value * scale))
        }
    }
}